README.md
# thesis-tool-viz-from-indent README

This is the README for your extension "thesis-tool-viz-from-indent". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: enable/disable this extension
* `myExtension.thing`: set to `blah` to do something

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

-----------------------------------------------------------------------------------------------------------

## Working with Markdown

**Note:** You can author your README using Visual Studio Code.  Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux)
* Toggle preview (`Shift+CMD+V` on macOS or `Shift+Ctrl+V` on Windows and Linux)
* Press `Ctrl+Space` (Windows, Linux) or `Cmd+Space` (macOS) to see a list of Markdown snippets

### For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "thesis-tool-viz-from-indent" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

out/decorate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decorate = void 0;
const vscode = require("vscode");
// from decoration tut
const functionDecorationType = vscode.window.createTextEditorDecorationType({
    backgroundColor: 'rgba(255, 0, 0, 0.2)',
    isWholeLine: true
});
const ifDecorationType = vscode.window.createTextEditorDecorationType({
    backgroundColor: 'rgba(0, 255, 0, 0.2)',
    isWholeLine: true
});
const functionDecorations = [];
const ifDecorations = [];
function decorate(editor) {
    let code = editor.document.getText();
    const document = editor.document;
    vscode.commands.executeCommand('vscode.executeDocumentSymbolProvider', document.uri).then(symbols => {
        symbols === null || symbols === void 0 ? void 0 : symbols.forEach(symbol => {
            decorateSymbolRecursive(symbol);
        });
    });
    editor.setDecorations(functionDecorationType, functionDecorations);
    editor.setDecorations(ifDecorationType, ifDecorations);
}
exports.decorate = decorate;
function decorateSymbolRecursive(symbol) {
    let range = symbol.range;
    let decoration = { range };
    switch (symbol.kind) {
        case vscode.SymbolKind.Function:
            functionDecorations.push(decoration);
            break;
        default:
            break;
    }
    symbol.children.forEach(child => {
        decorateSymbolRecursive(child);
    });
}
//# sourceMappingURL=decorate.js.map

src/decorate.ts
import * as vscode from "vscode";


// from decoration tut
const functionDecorationType = vscode.window.createTextEditorDecorationType({
    backgroundColor: 'rgba(255, 0, 0, 0.2)',
    isWholeLine: true
})
const ifDecorationType = vscode.window.createTextEditorDecorationType({
    backgroundColor: 'rgba(0, 255, 0, 0.2)',
    isWholeLine: true
})

const functionDecorations: vscode.DecorationOptions[] = [];
const ifDecorations: vscode.DecorationOptions[] = [];

export function decorate(editor: vscode.TextEditor): void {
    let code = editor.document.getText();

    const document = editor.document;

    vscode.commands.executeCommand<vscode.DocumentSymbol[]>('vscode.executeDocumentSymbolProvider', document.uri).then(symbols => {
        symbols?.forEach(symbol => {
            decorateSymbolRecursive(symbol);
        });
    });


    editor.setDecorations(functionDecorationType, functionDecorations);
    editor.setDecorations(ifDecorationType, ifDecorations);
}

function decorateSymbolRecursive(symbol: vscode.DocumentSymbol) {
    let range = symbol.range;
    let decoration = { range };

    switch (symbol.kind) {
        case vscode.SymbolKind.Function:
            functionDecorations.push(decoration);
            break;
        default:
            break;
    }

    symbol.children.forEach(child => {
        decorateSymbolRecursive(child);
    });
}

out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
const vscode = require("vscode");
const indentDecorator_1 = require("./indentDecorator");
// from decoration tut
const decorationType = vscode.window.createTextEditorDecorationType({
    backgroundColor: 'green',
    border: '2px solid white',
});
function activate(context) {
    console.log("Congratulations, your extension is now active!");
    vscode.workspace.onWillSaveTextDocument(event => {
        const openEditor = vscode.window.visibleTextEditors.filter(editor => editor.document.uri === event.document.uri)[0];
        // decorate(openEditor);
        indentDecorator_1.decorateIndent(openEditor);
    });
}
exports.activate = activate;
function write(code) {
    const editor = vscode.window.activeTextEditor;
    if (!editor) {
        throw new Error("There's no active editor");
    }
    const edit = new vscode.WorkspaceEdit();
    const wholeDocument = new vscode.Range(new vscode.Position(0, 0), new vscode.Position(editor.document.lineCount, 0));
    const updateCode = new vscode.TextEdit(wholeDocument, code);
    edit.set(editor.document.uri, [updateCode]);
    vscode.workspace.applyEdit(edit);
}
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map

src/extension.ts
import * as vscode from "vscode";
import { decorate } from "./decorate";
import { decorateIndent } from "./indentDecorator";

// from decoration tut
const decorationType = vscode.window.createTextEditorDecorationType({
	backgroundColor: 'green',
	border: '2px solid white',
})

export function activate(context: vscode.ExtensionContext) {
	console.log("Congratulations, your extension is now active!");

	vscode.workspace.onWillSaveTextDocument(event => {
		const openEditor = vscode.window.visibleTextEditors.filter(
			editor => editor.document.uri === event.document.uri
		)[0];
		// decorate(openEditor);
		decorateIndent(openEditor);
	});
}

function write(code: string): void {
	const editor = vscode.window.activeTextEditor;
	if (!editor) {
		throw new Error("There's no active editor");
	}

	const edit = new vscode.WorkspaceEdit();

	const wholeDocument = new vscode.Range(
		new vscode.Position(0, 0),
		new vscode.Position(editor.document.lineCount, 0)
	);
	const updateCode = new vscode.TextEdit(wholeDocument, code);

	edit.set(editor.document.uri, [updateCode]);

	vscode.workspace.applyEdit(edit);
}

export function deactivate() { }

src/indentDecorator.ts
import * as vscode from "vscode";

const decorations: any[] = [];
const decorationRanges: any[] = [];

enum DecorationPurpose {
    TITLE,
    END,
    CODE,
    INDENT
}

enum CodeBlockType {
    function,
    class,
    other
}

class CodeBlock {
    range: vscode.Range;
    type: CodeBlockType;
    indentLevel: number;
    constructor(range: vscode.Range, type: CodeBlockType, indentLevel: number) {
        this.range = range;
        this.type = type;
        this.indentLevel = indentLevel;
    }

    decorate() {
        // if (this.type != CodeBlockType.function) return;

        const tabsize = vscode.workspace.getConfiguration('editor').get<number>('tabSize') ?? 4;
        const useSpaces = vscode.workspace.getConfiguration('editor').get<boolean>('insertSpaces') ?? true;

        for (let line = this.range.start.line + 1; line <= this.range.end.line; ++line) {
            const startChar = useSpaces ? tabsize * this.indentLevel : this.indentLevel;
            const endChar = useSpaces ? startChar + tabsize : startChar + 1;
            const range = new vscode.Range(line, startChar, line, endChar);
            addDecoration(this.indentLevel, this.type, range, DecorationPurpose.INDENT);
        }
        const titleStart = (this.indentLevel) * tabsize;
        const titleEnd = 999;
        const titleRange = new vscode.Range(this.range.start.line, titleStart, this.range.start.line, titleEnd);
        addDecoration(this.indentLevel, this.type, titleRange, DecorationPurpose.TITLE);

        const endStart = (this.indentLevel) * tabsize;
        const endEnd = 999;
        const endRange = new vscode.Range(this.range.end.line, endStart, this.range.end.line, endEnd);
        addDecoration(this.indentLevel, this.type, endRange, DecorationPurpose.END);
    }
}

const codeblocks: CodeBlock[] = [];

export function decorateIndent(editor: vscode.TextEditor): void {
    // get the current indentation method, as defined by the settings
    vscode.workspace.getConfiguration().update('editor.lineHeight', 25, vscode.ConfigurationTarget.Workspace);
    vscode.workspace.getConfiguration().update('editor.formatOnType', true, vscode.ConfigurationTarget.Workspace);
    vscode.workspace.getConfiguration().update('editor.formatOnSave', true, vscode.ConfigurationTarget.Workspace);

    // vscode.commands.executeCommand('editor.action.formatDocument');

    const useSpaces = vscode.workspace.getConfiguration('editor').get<boolean>('insertSpaces') ?? true;
    useSpaces ? vscode.commands.executeCommand('editor.action.indentationToSpaces')
        : vscode.commands.executeCommand('editor.action.indentationToTabs');

    fillEmptyLines(editor);

    vscode.commands.executeCommand<vscode.DocumentSymbol[]>('vscode.executeDocumentSymbolProvider', editor.document.uri).then(symbols => {
        decorationRanges.forEach(options => {
            options[DecorationPurpose.TITLE].length = 0;
            options[DecorationPurpose.END].length = 0;
            options[DecorationPurpose.CODE].length = 0;
            options[DecorationPurpose.INDENT].length = 0;
        })

        codeblocks.length = 0;

        findRanges(editor);
        if (symbols)
            refineRangesBySymbols(editor, symbols);

        codeblocks.forEach(cb => cb.decorate());


        decorationRanges.forEach((options, index) => {
            if (options != null) {
                editor.setDecorations(decorations[index][DecorationPurpose.TITLE], options[DecorationPurpose.TITLE]);
                editor.setDecorations(decorations[index][DecorationPurpose.END], options[DecorationPurpose.END]);
                editor.setDecorations(decorations[index][DecorationPurpose.INDENT], options[DecorationPurpose.INDENT]);
                editor.setDecorations(decorations[index][DecorationPurpose.CODE], options[DecorationPurpose.CODE]);
            }
        })
    });
}

function findRanges(editor: vscode.TextEditor) {
    // get the current indentation method, as defined by the settings
    const tabsize = vscode.workspace.getConfiguration('editor').get<number>('tabSize') ?? 4;
    const useSpaces = vscode.workspace.getConfiguration('editor').get<boolean>('insertSpaces') ?? true;

    const openRanges: vscode.Position[] = [];
    // const closedRanges: vscode.Range[] = [];

    let indentLevel = -1;
    let lineNum = -1;
    const lines = editor.document.getText().split('\n');
    lines.forEach(line => {
        lineNum++;
        let amountOfWhitespaces = line.search(/\S|$/);
        let newIndentLevel = useSpaces ? Math.floor(amountOfWhitespaces / tabsize) : amountOfWhitespaces;
        let indentDelta = line.trim().length !== 0 && indentLevel >= 0 ? newIndentLevel - indentLevel : 0;
        indentLevel = line.trim().length !== 0 ? newIndentLevel : indentLevel;

        // discard indent level 0
        if (indentLevel === 0 && indentDelta >= 0) {
            return;
        }

        if (indentDelta > 0) {
            openRanges.push(new vscode.Position(lineNum - 1, 0));
        }
        if (indentDelta < 0) {
            const start = openRanges.pop();
            const end = new vscode.Position(lineNum, 999);
            if (start && end) {
                // closedRanges.push(new vscode.Range(start, end));
                let codeBlock = new CodeBlock(new vscode.Range(start, end), CodeBlockType.other, indentLevel);
                codeblocks.push(codeBlock);
            }
        }
    });

}

function refineRangesBySymbols(editor: vscode.TextEditor, symbols: vscode.DocumentSymbol[]) {
    symbols?.forEach(symbol => {
        refineRangesBySymbolsRecursive(editor, symbol);
    });
}

function refineRangesBySymbolsRecursive(editor: vscode.TextEditor, symbol: vscode.DocumentSymbol) {
    switch (symbol.kind) {
        case vscode.SymbolKind.Function:
            refineRange(symbol.range, CodeBlockType.function);
            break;
        default:
            break;
    }

    symbol.children.forEach(child => {
        refineRangesBySymbolsRecursive(editor, child);
    });
}

function refineRange(range: vscode.Range, type: CodeBlockType) {
    const codeblock = codeblocks.find(cb => {
        return cb.range.start.line == range.start.line && cb.range.end.line == range.end.line;
    });
    if (codeblock)
        codeblock.type = type;
}

function fillEmptyLines(editor: vscode.TextEditor) {
    // get the current indentation method, as defined by the settings
    const tabsize = vscode.workspace.getConfiguration('editor').get<number>('tabSize') ?? 4;
    const useSpaces = vscode.workspace.getConfiguration('editor').get<boolean>('insertSpaces') ?? true;
    vscode.workspace.getConfiguration().update('editor.lineHeight', 25, vscode.ConfigurationTarget.Workspace);

    useSpaces ? vscode.commands.executeCommand('editor.action.indentationToSpaces')
        : vscode.commands.executeCommand('editor.action.indentationToTabs');


    let indentLevel = -1;
    let lineNum = -1;
    const lines = editor.document.getText().split('\n');
    editor.edit(edit => {
        lines.forEach(line => {
            lineNum++;
            const amountOfWhitespaces = line.search(/\S|$/);
            const newIndentLevel = useSpaces ? Math.floor(amountOfWhitespaces / tabsize) : amountOfWhitespaces;
            indentLevel = line.trim().length !== 0 ? newIndentLevel : indentLevel;

            // add indentation to empty lines
            if (line.trim().length === 0) {
                const indentSpaces = '' + (useSpaces ? ' '.repeat(tabsize * indentLevel) : '\t'.repeat(indentLevel));
                edit.replace(new vscode.Range(lineNum, 0, lineNum, 999), indentSpaces);
            }
        })
    });
}

function addDecoration(indentLevel: number, type: CodeBlockType, range: vscode.Range, purpose: DecorationPurpose) {
    let decoration = { range };

    if (decorationRanges[type] == null) {
        decorationRanges[type] = createDecorationOption();
    }
    if (decorations[type] == null) {
        decorations[type] = createDecorationType(type);
    }

    decorationRanges[type][purpose].push(decoration);
}

function createDecorationOption() {
    const titleDecOptions: vscode.DecorationOptions[] = [];
    const endDecOptions: vscode.DecorationOptions[] = [];
    const indentDecOptions: vscode.DecorationOptions[] = [];
    const codeDecOptions: vscode.DecorationOptions[] = [];

    return {
        [DecorationPurpose.TITLE]: titleDecOptions,
        [DecorationPurpose.END]: endDecOptions,
        [DecorationPurpose.INDENT]: indentDecOptions,
        [DecorationPurpose.CODE]: codeDecOptions
    }
}

function createDecorationType(type: CodeBlockType) {

    const typeColors = {
        [CodeBlockType.class]: '255, 100, 100',
        [CodeBlockType.function]: '100, 255, 100',
        [CodeBlockType.other]: '100, 100, 255'
    }

    const color = typeColors[type];

    const titleDecType = vscode.window.createTextEditorDecorationType({
        backgroundColor: `rgba(${color}, 0.5)`,
        outline: `2px solid rgba(${color}, 1)`,
        before: {
            backgroundColor: `rgba(${color}, 1)`,
            width: '2px',
            height: '50%',
            contentText: '',
            margin: '-100% 0 0 0'
        }
    })
    const endDecType = vscode.window.createTextEditorDecorationType({
        backgroundColor: `rgba(${color}, 0.5)`,
        outline: `2px solid rgba(${color}, 1)`,
    })
    const codeDecType = vscode.window.createTextEditorDecorationType({
        // backgroundColor: 'rgba(255, 200, 200, 0.5)',
    })
    const indentDecType = vscode.window.createTextEditorDecorationType({
        backgroundColor: `rgba(${color}, 0.25)`,
        before: {
            backgroundColor: `rgba(${color}, 1)`,
            width: '2px',
            height: '100%',
            contentText: '',
            margin: '-100% 0 0 0'
        }
    })

    return {
        [DecorationPurpose.TITLE]: titleDecType,
        [DecorationPurpose.END]: endDecType,
        [DecorationPurpose.INDENT]: indentDecType,
        [DecorationPurpose.CODE]: codeDecType
    }
}

out/indentDecorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decorateIndent = void 0;
const vscode = require("vscode");
const decorations = [];
const decorationRanges = [];
var DecorationPurpose;
(function (DecorationPurpose) {
    DecorationPurpose[DecorationPurpose["TITLE"] = 0] = "TITLE";
    DecorationPurpose[DecorationPurpose["END"] = 1] = "END";
    DecorationPurpose[DecorationPurpose["CODE"] = 2] = "CODE";
    DecorationPurpose[DecorationPurpose["INDENT"] = 3] = "INDENT";
})(DecorationPurpose || (DecorationPurpose = {}));
var CodeBlockType;
(function (CodeBlockType) {
    CodeBlockType[CodeBlockType["function"] = 0] = "function";
    CodeBlockType[CodeBlockType["class"] = 1] = "class";
    CodeBlockType[CodeBlockType["other"] = 2] = "other";
})(CodeBlockType || (CodeBlockType = {}));
class CodeBlock {
    constructor(range, type, indentLevel) {
        this.range = range;
        this.type = type;
        this.indentLevel = indentLevel;
    }
    decorate() {
        // if (this.type != CodeBlockType.function) return;
        var _a, _b;
        const tabsize = (_a = vscode.workspace.getConfiguration('editor').get('tabSize')) !== null && _a !== void 0 ? _a : 4;
        const useSpaces = (_b = vscode.workspace.getConfiguration('editor').get('insertSpaces')) !== null && _b !== void 0 ? _b : true;
        for (let line = this.range.start.line + 1; line <= this.range.end.line; ++line) {
            const startChar = useSpaces ? tabsize * this.indentLevel : this.indentLevel;
            const endChar = useSpaces ? startChar + tabsize : startChar + 1;
            const range = new vscode.Range(line, startChar, line, endChar);
            addDecoration(this.indentLevel, this.type, range, DecorationPurpose.INDENT);
        }
        const titleStart = (this.indentLevel) * tabsize;
        const titleEnd = 999;
        const titleRange = new vscode.Range(this.range.start.line, titleStart, this.range.start.line, titleEnd);
        addDecoration(this.indentLevel, this.type, titleRange, DecorationPurpose.TITLE);
        const endStart = (this.indentLevel) * tabsize;
        const endEnd = 999;
        const endRange = new vscode.Range(this.range.end.line, endStart, this.range.end.line, endEnd);
        addDecoration(this.indentLevel, this.type, endRange, DecorationPurpose.END);
    }
}
const codeblocks = [];
function decorateIndent(editor) {
    var _a;
    // get the current indentation method, as defined by the settings
    vscode.workspace.getConfiguration().update('editor.lineHeight', 25, vscode.ConfigurationTarget.Workspace);
    vscode.workspace.getConfiguration().update('editor.formatOnType', true, vscode.ConfigurationTarget.Workspace);
    vscode.workspace.getConfiguration().update('editor.formatOnSave', true, vscode.ConfigurationTarget.Workspace);
    // vscode.commands.executeCommand('editor.action.formatDocument');
    const useSpaces = (_a = vscode.workspace.getConfiguration('editor').get('insertSpaces')) !== null && _a !== void 0 ? _a : true;
    useSpaces ? vscode.commands.executeCommand('editor.action.indentationToSpaces')
        : vscode.commands.executeCommand('editor.action.indentationToTabs');
    fillEmptyLines(editor);
    vscode.commands.executeCommand('vscode.executeDocumentSymbolProvider', editor.document.uri).then(symbols => {
        decorationRanges.forEach(options => {
            options[DecorationPurpose.TITLE].length = 0;
            options[DecorationPurpose.END].length = 0;
            options[DecorationPurpose.CODE].length = 0;
            options[DecorationPurpose.INDENT].length = 0;
        });
        codeblocks.length = 0;
        findRanges(editor);
        if (symbols)
            refineRangesBySymbols(editor, symbols);
        codeblocks.forEach(cb => cb.decorate());
        decorationRanges.forEach((options, index) => {
            if (options != null) {
                editor.setDecorations(decorations[index][DecorationPurpose.TITLE], options[DecorationPurpose.TITLE]);
                editor.setDecorations(decorations[index][DecorationPurpose.END], options[DecorationPurpose.END]);
                editor.setDecorations(decorations[index][DecorationPurpose.INDENT], options[DecorationPurpose.INDENT]);
                editor.setDecorations(decorations[index][DecorationPurpose.CODE], options[DecorationPurpose.CODE]);
            }
        });
    });
}
exports.decorateIndent = decorateIndent;
function findRanges(editor) {
    var _a, _b;
    // get the current indentation method, as defined by the settings
    const tabsize = (_a = vscode.workspace.getConfiguration('editor').get('tabSize')) !== null && _a !== void 0 ? _a : 4;
    const useSpaces = (_b = vscode.workspace.getConfiguration('editor').get('insertSpaces')) !== null && _b !== void 0 ? _b : true;
    const openRanges = [];
    // const closedRanges: vscode.Range[] = [];
    let indentLevel = -1;
    let lineNum = -1;
    const lines = editor.document.getText().split('\n');
    lines.forEach(line => {
        lineNum++;
        let amountOfWhitespaces = line.search(/\S|$/);
        let newIndentLevel = useSpaces ? Math.floor(amountOfWhitespaces / tabsize) : amountOfWhitespaces;
        let indentDelta = line.trim().length !== 0 && indentLevel >= 0 ? newIndentLevel - indentLevel : 0;
        indentLevel = line.trim().length !== 0 ? newIndentLevel : indentLevel;
        // discard indent level 0
        if (indentLevel === 0 && indentDelta >= 0) {
            return;
        }
        if (indentDelta > 0) {
            openRanges.push(new vscode.Position(lineNum - 1, 0));
        }
        if (indentDelta < 0) {
            const start = openRanges.pop();
            const end = new vscode.Position(lineNum, 999);
            if (start && end) {
                // closedRanges.push(new vscode.Range(start, end));
                let codeBlock = new CodeBlock(new vscode.Range(start, end), CodeBlockType.other, indentLevel);
                codeblocks.push(codeBlock);
            }
        }
    });
}
function refineRangesBySymbols(editor, symbols) {
    symbols === null || symbols === void 0 ? void 0 : symbols.forEach(symbol => {
        refineRangesBySymbolsRecursive(editor, symbol);
    });
}
function refineRangesBySymbolsRecursive(editor, symbol) {
    switch (symbol.kind) {
        case vscode.SymbolKind.Function:
            refineRange(symbol.range, CodeBlockType.function);
            break;
        default:
            break;
    }
    symbol.children.forEach(child => {
        refineRangesBySymbolsRecursive(editor, child);
    });
}
function refineRange(range, type) {
    const codeblock = codeblocks.find(cb => {
        return cb.range.start.line == range.start.line && cb.range.end.line == range.end.line;
    });
    if (codeblock)
        codeblock.type = type;
}
function fillEmptyLines(editor) {
    var _a, _b;
    // get the current indentation method, as defined by the settings
    const tabsize = (_a = vscode.workspace.getConfiguration('editor').get('tabSize')) !== null && _a !== void 0 ? _a : 4;
    const useSpaces = (_b = vscode.workspace.getConfiguration('editor').get('insertSpaces')) !== null && _b !== void 0 ? _b : true;
    vscode.workspace.getConfiguration().update('editor.lineHeight', 25, vscode.ConfigurationTarget.Workspace);
    useSpaces ? vscode.commands.executeCommand('editor.action.indentationToSpaces')
        : vscode.commands.executeCommand('editor.action.indentationToTabs');
    let indentLevel = -1;
    let lineNum = -1;
    const lines = editor.document.getText().split('\n');
    editor.edit(edit => {
        lines.forEach(line => {
            lineNum++;
            const amountOfWhitespaces = line.search(/\S|$/);
            const newIndentLevel = useSpaces ? Math.floor(amountOfWhitespaces / tabsize) : amountOfWhitespaces;
            indentLevel = line.trim().length !== 0 ? newIndentLevel : indentLevel;
            // add indentation to empty lines
            if (line.trim().length === 0) {
                const indentSpaces = '' + (useSpaces ? ' '.repeat(tabsize * indentLevel) : '\t'.repeat(indentLevel));
                edit.replace(new vscode.Range(lineNum, 0, lineNum, 999), indentSpaces);
            }
        });
    });
}
function addDecoration(indentLevel, type, range, purpose) {
    let decoration = { range };
    if (decorationRanges[type] == null) {
        decorationRanges[type] = createDecorationOption();
    }
    if (decorations[type] == null) {
        decorations[type] = createDecorationType(type);
    }
    decorationRanges[type][purpose].push(decoration);
}
function createDecorationOption() {
    const titleDecOptions = [];
    const endDecOptions = [];
    const indentDecOptions = [];
    const codeDecOptions = [];
    return {
        [DecorationPurpose.TITLE]: titleDecOptions,
        [DecorationPurpose.END]: endDecOptions,
        [DecorationPurpose.INDENT]: indentDecOptions,
        [DecorationPurpose.CODE]: codeDecOptions
    };
}
function createDecorationType(type) {
    const typeColors = {
        [CodeBlockType.class]: '255, 100, 100',
        [CodeBlockType.function]: '100, 255, 100',
        [CodeBlockType.other]: '100, 100, 255'
    };
    const color = typeColors[type];
    const titleDecType = vscode.window.createTextEditorDecorationType({
        backgroundColor: `rgba(${color}, 0.5)`,
        outline: `2px solid rgba(${color}, 1)`,
        before: {
            backgroundColor: `rgba(${color}, 1)`,
            width: '2px',
            height: '50%',
            contentText: '',
            margin: '-100% 0 0 0'
        }
    });
    const endDecType = vscode.window.createTextEditorDecorationType({
        backgroundColor: `rgba(${color}, 0.5)`,
        outline: `2px solid rgba(${color}, 1)`,
    });
    const codeDecType = vscode.window.createTextEditorDecorationType({
    // backgroundColor: 'rgba(255, 200, 200, 0.5)',
    });
    const indentDecType = vscode.window.createTextEditorDecorationType({
        backgroundColor: `rgba(${color}, 0.25)`,
        before: {
            backgroundColor: `rgba(${color}, 1)`,
            width: '2px',
            height: '100%',
            contentText: '',
            margin: '-100% 0 0 0'
        }
    });
    return {
        [DecorationPurpose.TITLE]: titleDecType,
        [DecorationPurpose.END]: endDecType,
        [DecorationPurpose.INDENT]: indentDecType,
        [DecorationPurpose.CODE]: codeDecType
    };
}
//# sourceMappingURL=indentDecorator.js.map